### PR TITLE
[Discussion] Interfaces

### PR DESCRIPTION
**Pull request by stegben**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,15 +3,39 @@
...
+p.add(step())
+
+p.fit()
+p.transform()
+
+p.inverse_transform()
+```
+
+### Serialization
+```pythn
```
Comment (absolutelyNoWarranty) [inline README.md]: python

File: strpipe/steps.pyx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        self.op = op_factory[op_name]
+
+    def fit(self, strs):
+        self.state = self.op.fit(strs)
+
+    def transform(self, strs):
+        self.op.transform(self.state, strs)
+
+    def inverse_transform(self, strs):
+        pass
```
Comment (SoluMilken) [inline strpipe/steps.pyx]: self.op.inverse_transform

**Comment by SoluMilken** [comment]
崩崩 我 git push -f 惹

**Comment by stegben** [comment]
@SoluMilken @absolutelyNoWarranty  @noobOriented 
How is this usage looks like? Would like to know your opinion:
```python
import strpipe as sp

p = sp.Pipe()
p.add_step_by_op_name(
    op_name='Trim',
    op_kwargs={'tokens': ['\n', '\r']},
)
p.add_step_by_op_name('CharTokenize')
p.add_step_by_op_name(
    op_name='MapStringToIndex',
    default_state={'你': 0, '好': 1, '早': 2},
    fixed_state=True,  # the state of the step won't change even `p.fit` is called
)

data = sp.TextData([
    '你好啊\n',
    '早安',
    '你早上好\n',
])

p.fit(data)
result, tx_info = p.transform(data)  # convention: tx => tranform
back_data = p.inverse_transform(result, tx_info)
```

**Comment by SoluMilken** [comment]
What will happen if 
(1) I gave default state and set fixed_state = False
(2) do not provide default state, but fixed_state = True 
(3) do not provide default state and fixed_state = False

**Comment by SoluMilken** [comment]
MapStringToIndex 的 反向版 的 state 是自動生成的嗎？ 
（任何一個op 的 state 都只能指定 順向, 反向的都是用正向的來生成???)

**Comment by stegben** [comment]
Nope, the state is only generated by the forward pass. The `inverse_transform` utilize this state to perform operation.

File: strpipe/ops/base.pyx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+cdef class BaseOp:
```
Comment (SoluMilken) [inline strpipe/ops/base.pyx]: # distutils: language = c++

Comment (stegben) [reply strpipe/ops/base.pyx]: done

Comment (SoluMilken) [reply strpipe/ops/base.pyx]: 喔喔喔喔喔

**Pull request merged**